libraries/AP_ICEngine/AP_ICEngine.cpp

```diff
@@ -402,8 +410,15 @@ void AP_ICEngine::update(void)
         if (!should_run) {
             state = ICE_OFF;
         } else if (now - starter_last_run_ms >= starter_delay*1000) {
-            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Starting engine");
-            state = ICE_STARTING;
+            // check if we should retry starting the engine
+            if (max_crank_retry <= 0 || crank_retry_ct < max_crank_retry) {
+                GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Starting engine");
+                state = ICE_STARTING;
+                crank_retry_ct++;
```

Comment:
I think we should reset this on a un-commanded stop. You could build up some retrys on initial start then run for some long period of time. Then the engine stops without being commanded and you only have one retry left. Here:
https://github.com/ArduPilot/ardupilot/blob/04ee66bec4f6c5aa8e2f0d21a3facd34f1f8f29b/libraries/AP_ICEngine/AP_ICEngine.cpp#L430

Reply:
This would take a larger refactor of ICE than we wanted to push right now. There's currently no difference in the code between an unsuccessful start attempt and an uncommanded engine stop. Both go through the "engine running" state.